2.0-proposal/readme.md

```diff
@@ -0,0 +1,183 @@
+# Introduction
+
+This folder contains the design for the charms.reactive 2.0 reboot.
+
+*Tip: Forget about hooks!*
+
+# Index
+
+[VISION.md](VISION.md) offers a short description of what charms.reactive is about and where we want to go.
+
+# In general
+
+Events trigger (set or unset) flags. Handlers react to flags and change them. Flags can have data attached to them.
+
+<!-- triggers are suggested here: https://github.com/juju-solutions/charms.reactive/issues/97 -->
+
+That's it.
+
+# Types of Events
+
+### The world around us has changed! [external]
+
+The external state outside of this unit has changed and we need to react to that.
+
+Ex: `relation.changed` and `config.changed`.
+
+### Lifecycle transition requested! [external]
+
+The operator has requested that we bring the application into a new state of its lifecycle.
+
+Ex: `install`, `start`, `stop`.
+
+<!-- hooks are just one of the many types of events, decoupling the reactive bus from hooks as proposed here: https://github.com/juju-solutions/charms.reactive/issues/90 -->
+
+### Our internal state has changed! [internal]
+
+The internal state of the unit has changed and we need to react to that.
+
+Ex: A file changing (`file.changed.<xyz>`), landscape messing about on the system (removal of `apt.installed.xyz`), a flag being toggled or the data of a flag being changed (`config.changed` unsets `config.checked`).
+
+<!-- Actions is a tricky one. I'd wait a bit to include actions in the story until we're more certain of where we want to go.
+
+### Action requested! [external]
+
+The operator has requested an action that shouldn't significantly impact the state of the application.
+
+Ex: `smoke-test`, `run-job`, `create-backup`, ... -->
+
+# When do triggers and handlers run?
+
+The flow of the framework is the following:
+
+1. The reactive framework will run a handler when all its preconditions are met. A handler that has run is inactive. It will not run again until one of the flags it reacts to has changed. This change can be either the flag being toggled, or the flag data changing.
+
+2. After each handler run, all internal events will be processed. The reactive framework will not remove flags itself. Only triggers and handlers can do that.
```

Comment:
@galgalesh "I'm proposing to never rerun the handler if the preconditions stay true" - then they wouldn't be preconditions anymore. I feel like this breaks the nice clean model we have right now. What about the current implementation irks you exactly? I'm not sure why you would want to do something like this (no offense)? To me, (possibly just on first take) this seems to make things more unclear/limiting to the user, then the current implementation. Are there others that want this new implementation also?